ec/ecresource/deploymentresource/apm/v2/apm.go

```diff
@@ -0,0 +1,199 @@
+	HttpEndpoint              *string    `tfsdk:"http_endpoint"`
+	HttpsEndpoint             *string    `tfsdk:"https_endpoint"`
+	InstanceConfigurationId   *string    `tfsdk:"instance_configuration_id"`
+	Size                      *string    `tfsdk:"size"`
+	SizeResource              *string    `tfsdk:"size_resource"`
+	ZoneCount                 int        `tfsdk:"zone_count"`
+	Config                    *ApmConfig `tfsdk:"config"`
+}
+
+func ReadApms(in []*models.ApmResourceInfo) (*Apm, error) {
+	for _, model := range in {
+		if util.IsCurrentApmPlanEmpty(model) || utils.IsApmResourceStopped(model) {
+			continue
+		}
+
+		apm, err := ReadApm(model)
+		if err != nil {
+			return nil, err
+		}
+
+		return apm, nil
+	}
+
+	return nil, nil
+}
+
+func ReadApm(in *models.ApmResourceInfo) (*Apm, error) {
+	var apm Apm
+
+	apm.RefId = in.RefID
+
+	apm.ResourceId = in.Info.ID
+
+	apm.Region = in.Region
+
+	plan := in.Info.PlanInfo.Current.Plan
+
+	topologies, err := ReadApmTopologies(plan.ClusterTopology)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(topologies) > 0 {
+		apm.InstanceConfigurationId = topologies[0].InstanceConfigurationId
+		apm.Size = topologies[0].Size
+		apm.SizeResource = topologies[0].SizeResource
+		apm.ZoneCount = topologies[0].ZoneCount
+	}
+
+	apm.ElasticsearchClusterRefId = in.ElasticsearchClusterRefID
+
+	apm.HttpEndpoint, apm.HttpsEndpoint = converters.ExtractEndpoints(in.Info.Metadata)
+
+	configs, err := readApmConfigs(plan.Apm)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(configs) > 0 {
+		apm.Config = &configs[0]
+	}
+
+	return &apm, nil
+}
+
+func (apm ApmTF) Payload(ctx context.Context, payload models.ApmPayload) (*models.ApmPayload, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if !apm.ElasticsearchClusterRefId.IsNull() {
+		payload.ElasticsearchClusterRefID = &apm.ElasticsearchClusterRefId.Value
+	}
+
+	if !apm.RefId.IsNull() {
+		payload.RefID = &apm.RefId.Value
+	}
+
+	if apm.Region.Value != "" {
+		payload.Region = &apm.Region.Value
+	}
+
+	if !apm.Config.IsNull() && !apm.Config.IsUnknown() {
+		var cfg v1.ApmConfigTF
+
+		ds := tfsdk.ValueAs(ctx, apm.Config, &cfg)
+
+		diags.Append(ds...)
+
+		if !ds.HasError() {
+			diags.Append(apmConfigPayload(ctx, cfg, payload.Plan.Apm)...)
+		}
+	}
+
+	topology := topologyv1.TopologyTF{
+		InstanceConfigurationId: apm.InstanceConfigurationId,
+		Size:                    apm.Size,
+		SizeResource:            apm.SizeResource,
+		ZoneCount:               apm.ZoneCount,
+	}
+
+	topologyPayload, ds := apmTopologyPayload(ctx, topology, defaultApmTopology(payload.Plan.ClusterTopology), 0)
+
+	diags.Append(ds...)
+
+	if !ds.HasError() && topologyPayload != nil {
+		payload.Plan.ClusterTopology = []*models.ApmTopologyElement{topologyPayload}
+	}
+
+	return &payload, diags
+}
+
+func ApmPayload(ctx context.Context, apmObj types.Object, template *models.DeploymentTemplateInfoV2) (*models.ApmPayload, diag.Diagnostics) {
```

Comment:
Similarly, this is `EnrichTemplate` or something

Reply:
From a caller perspective the func generates payload while `EnrichTemplate` describes more implementation details of the func.